src/main/java/com/okta/tools/OktaAwsCliAssumeRole.java

```diff
@@ -94,8 +97,23 @@ private static Path getSessionFilePath() {
 
     String run(Instant startInstant) throws Exception {
         Optional<Session> session = getCurrentSession();
-        if (session.isPresent() && sessionIsActive(startInstant, session.get()))
-            return session.get().profileName;
+        Instant expiry = startInstant.plus(864000 - 30, ChronoUnit.SECONDS);
+        String profileStore = System.getProperty("user.home") + "/.aws/.okta-profile-expiry";
+        Ini ini = new Ini(new File(profileStore));
+
+        if (session.isPresent() && sessionIsActive(startInstant, session.get())) {
+            if (OktaAwsConfig.createAwscli().oktaProfile != null) {
+                String profile = OktaAwsConfig.createAwscli().oktaProfile;
+                Set <String> activeSessions = ini.keySet();
+                if (!activeSessions.contains(profile)) {
+                    addOrUpdateProfile(OktaAwsConfig.createAwscli().oktaProfile, session.get().toString(), expiry);
+                }
+            }
+            else {
+                addOrUpdateProfile(session.get().profileName, "", expiry);
+            }
+                return session.get().profileName;
```

Comment:
:art: fix indentation.

Reply:
The logic for determining if there is an existing session that you can reuse needs to be separated from the high-level logic of the run method. The run method should read like a set of high-level steps and not get muddled in details like map-manipulation and session contains calls.

A key thing in refactoring this is to separate the code that checks or reads the state from the code that writes it back. It will be way more challenging to debug if the session reading and writing remain mixed like this.